src/arch/x86_64/mod.rs

```diff
@@ -111,101 +110,23 @@ pub fn get_cpu_frequency_from_os() -> std::result::Result<u32, FrequencyDetectio
-
-/// Creates the pagetables and the GDT in the guest memory space.
-///
-/// The memory slice must be larger than [`MIN_PHYSMEM_SIZE`].
-/// Also, the memory `mem` needs to be zeroed for [`PAGE_SIZE`] bytes at the
-/// offsets [`BOOT_PML4`] and [`BOOT_PDPTE`], otherwise the integrity of the
-/// pagetables and thus the integrity of the guest's memory is not ensured
-pub fn initialize_pagetables(mem: &mut [u8]) {
-	assert!(mem.len() >= MIN_PHYSMEM_SIZE);
-	let mem_addr = std::ptr::addr_of_mut!(mem[0]);
-
-	let (gdt_entry, pml4, pdpte, pde);
-	// Safety:
-	// We only operate in `mem`, which is plain bytes and we have ownership of
-	// these and it is asserted to be large enough.
-	unsafe {
-		gdt_entry = mem_addr
-			.add(BOOT_GDT.as_u64() as usize)
-			.cast::<[u64; 3]>()
-			.as_mut()
-			.unwrap();
-
-		pml4 = mem_addr
-			.add(BOOT_PML4.as_u64() as usize)
-			.cast::<PageTable>()
-			.as_mut()
-			.unwrap();
-		pdpte = mem_addr
-			.add(BOOT_PDPTE.as_u64() as usize)
-			.cast::<PageTable>()
-			.as_mut()
-			.unwrap();
-		pde = mem_addr
-			.add(BOOT_PDE.as_u64() as usize)
-			.cast::<PageTable>()
-			.as_mut()
-			.unwrap();
-
-		/* For simplicity we currently use 2MB pages and only a single
-		PML4/PDPTE/PDE. */
-
-		// per default is the memory zeroed, which we allocate by the system
-		// call mmap, so the following is not necessary:
-		/*libc::memset(pml4 as *mut _ as *mut libc::c_void, 0x00, PAGE_SIZE);
-		libc::memset(pdpte as *mut _ as *mut libc::c_void, 0x00, PAGE_SIZE);
-		libc::memset(pde as *mut _ as *mut libc::c_void, 0x00, PAGE_SIZE);*/
-	}
-	// initialize GDT
-	gdt_entry[BOOT_GDT_NULL] = 0;
-	gdt_entry[BOOT_GDT_CODE] = create_gdt_entry(0xA09B, 0, 0xFFFFF);
-	gdt_entry[BOOT_GDT_DATA] = create_gdt_entry(0xC093, 0, 0xFFFFF);
-
-	pml4[0].set_addr(
-		BOOT_PDPTE,
-		PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
-	);
-	pml4[511].set_addr(
-		BOOT_PML4,
-		PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
-	);
-	pdpte[0].set_addr(BOOT_PDE, PageTableFlags::PRESENT | PageTableFlags::WRITABLE);
-
-	for i in 0..512 {
-		let addr = PhysAddr::new(i as u64 * Page::<Size2MiB>::SIZE);
-		pde[i].set_addr(
-			addr,
-			PageTableFlags::PRESENT | PageTableFlags::WRITABLE | PageTableFlags::HUGE_PAGE,
-		);
-	}
-}
-
 /// Converts a virtual address in the guest to a physical address in the guest
 pub fn virt_to_phys(
 	addr: GuestVirtAddr,
 	mem: &MmapMemory,
-	pagetable_l0: GuestPhysAddr,
 ) -> Result<GuestPhysAddr, PagetableError> {
 	/// Number of Offset bits of a virtual address for a 4 KiB page, which are shifted away to get its Page Frame Number (PFN).
 	pub const PAGE_BITS: u64 = 12;
 
 	/// Number of bits of the index in each table (PML4, PDPT, PDT, PGT).
 	pub const PAGE_MAP_BITS: usize = 9;
 
-	let mut page_table =
-		unsafe { (mem.host_address(pagetable_l0).unwrap() as *mut PageTable).as_mut() }.unwrap();
+	let mut page_table = unsafe {
+		(mem.host_address(GuestPhysAddr::new(mem.guest_address.as_u64() + PML4_OFFSET))
```

Comment:
This will not work out, because of a planned switch from `MmapMemory` to `vm_memory::map::GuestRegionMmap`.

@jounathaen, @mkroening: Despite the fact that it's possible to put the `guest_address` value in a cell / container (`std::cell::OnceCell`, `std::sync::LazyLock`), there is still one big question remaining from me: Regardless of the type of the object representing the memory, the object is still stored in an `Arc<...>` - but we can only know whether the kernel is relocatable as soon as we are in *load_kernel* (`vm.rs`), how would we work around the fact that the memory is already "set" (therefore the `guest_address` cannot be changed without a setter, `cannot assign to data in an Arc`) and at a specific `guest_address` by the time we find out?

Reply:
Experimental `OnceLock` implementation, not complete because of the issue mentioned above:

- https://github.com/n0toose/uhyve/tree/aslr-oncelock-concept 
- https://github.com/hermit-os/uhyve/commit/1085671f6cdc04e805ba038a05629c408efc0d19 (based on https://github.com/hermit-os/uhyve/pull/725/commits/08bfcfd2caa2c5ed633cd0f7786188644be2d520)